File: android/app/src/main/python/stock_monitoring_app/indicators/breakout_indicator.py
import pandas as pd
from .base_indicator import Indicator

class BreakoutIndicator(Indicator):

    @staticmethod
    def get_search_space():
        return {
            "window": [10, 20, 30],
            # Add more parameters here if your indicator supports/needs them
        }

    """
    Identifies price breakouts above a recent high or below a recent low.
    """
    def __init__(self,
                 df: pd.DataFrame,
                 window: int = 20,
                 high_col: str = 'High',
                 low_col: str = 'Low',
                 close_col: str = 'Close'):
        """
        Args:
            df: OHLCV DataFrame.
            window: The lookback period to determine recent high/low.
            high_col: Name of the high price column.
            low_col: Name of the low price column.
            close_col: Name of the close price column.
        """

        super().__init__(df)
        self.window = window

        # Rolling max/min needs 'window' periods for first value, then shift(1).
        # So, to get at least one comparable value, len(df) >= window + 1.
        if len(self.df) < self.window + 1:
            raise ValueError(
                f"Insufficient data for {self.__class__.__name__} (window: {self.window}): "
                f"{len(self.df)} rows provided, requires at least {self.window + 1} rows."
            )
        self.high_col = high_col

        self.low_col = low_col
        self.close_col = close_col

        # Populate signal orientations (the attribute self.signal_orientations is initialized in super().__init__())
        # self.signal_orientations: dict[str, str] = {} # This line was redundant and has been removed.
        # Ensure these names match the columns generated in the calculate() method
        bullish_signal_col_name = f'Breakout_Bullish_Signal_{self.window}'
        bearish_signal_col_name = f'Breakout_Bearish_Signal_{self.window}'        
        self.signal_orientations[bullish_signal_col_name] = 'buy' # Bullish breakout is a buy signal
        self.signal_orientations[bearish_signal_col_name] = 'sell' # Bearish breakout is a sell signal

        # Check for required columns
        for col in [high_col, low_col, close_col]:
            if col not in self.df.columns:
                raise ValueError(f"Column '{col}' not found in DataFrame for Breakout calculation.")

    def calculate(self) -> pd.DataFrame:
        """
        Adds 'Breakout_Bullish_Signal_<window>' and 'Breakout_Bearish_Signal_<window>' columns.
        - Breakout_Bullish_Signal: True for bullish breakout, False otherwise.
        - Breakout_Bearish_Signal: True for bearish breakout, False otherwise.
        """
        bullish_signal_col_name = f'Breakout_Bullish_Signal_{self.window}'
        bearish_signal_col_name = f'Breakout_Bearish_Signal_{self.window}'

        # Calculate recent high and low (excluding current bar, so shift(1))
        recent_high = self.df[self.high_col].rolling(window=self.window, min_periods=self.window).max().shift(1)
        recent_low = self.df[self.low_col].rolling(window=self.window, min_periods=self.window).min().shift(1)

        # Bullish breakout: Close breaks above recent N-period high
        bullish_breakout_condition = (self.df[self.close_col] > recent_high)
        self.df[bullish_signal_col_name] = bullish_breakout_condition.fillna(False).astype(bool)

        # Bearish breakout: Close breaks below recent N-period low
        bearish_breakout_condition = (self.df[self.close_col] < recent_low)
        self.df[bearish_signal_col_name] = bearish_breakout_condition.fillna(False).astype(bool)

        return self.df

File: src/breakout_indicator.cpp
#include "breakout_indicator.hpp"
#include <algorithm>
#include <cmath>
#include <stdexcept>

// Utility: get reference to a column by name
static const std::vector<double>& get_column(const DataFrame& df, const std::string& name) {
    if (name == "High") return df.high;
    if (name == "Low") return df.low;
    if (name == "Close") return df.close;
    if (name == "Open") return df.open;
    if (name == "Volume") return df.volume;
    auto it = df.extra_columns.find(name);
    if (it != df.extra_columns.end()) return it->second;
    throw std::invalid_argument("Column not found: " + name);
}

BreakoutIndicator::BreakoutIndicator(const DataFrame& df, int window,
                                     const std::string& high_col,
                                     const std::string& low_col,
                                     const std::string& close_col)
    : Indicator(df), window_(window),
      high_col_(high_col), low_col_(low_col), close_col_(close_col)
{
    // No need to check .columns, just check the actual members
    size_t n = get_column(df_, close_col_).size();
    if (n < static_cast<size_t>(window_) + 1) {
        throw std::invalid_argument("Insufficient data for BreakoutIndicator (window: " +
            std::to_string(window_) + "): " +
            std::to_string(n) +
            " rows provided, requires at least " + std::to_string(window_ + 1) + " rows.");
    }
    bullish_signal_col_ = "Breakout_Bullish_Signal_" + std::to_string(window_);
    bearish_signal_col_ = "Breakout_Bearish_Signal_" + std::to_string(window_);

    signal_orientations_[bullish_signal_col_] = "buy";
    signal_orientations_[bearish_signal_col_] = "sell";
}

DataFrame BreakoutIndicator::calculate() {
    const auto& high = get_column(df_, high_col_);
    const auto& low = get_column(df_, low_col_);
    const auto& close = get_column(df_, close_col_);
    size_t n = close.size();

    std::vector<bool> bullish_signal(n, false);
    std::vector<bool> bearish_signal(n, false);

    for (size_t i = window_; i < n; ++i) {
        double recent_high = *std::max_element(high.begin() + (i - window_), high.begin() + i);
        double recent_low = *std::min_element(low.begin() + (i - window_), low.begin() + i);
        if (close[i] > recent_high)
            bullish_signal[i] = true;
        if (close[i] < recent_low)
            bearish_signal[i] = true;
    }

    DataFrame out = df_;
    out.extra_bool_columns[bullish_signal_col_] = bullish_signal;
    out.extra_bool_columns[bearish_signal_col_] = bearish_signal;
    return out;
}
